vulnerabilities/models.py

```diff
@@ -1098,3 +1149,192 @@ class ApiUser(UserModel):
+        )
+
+
+class VulnerabilityChangeLog(ChangeLog):
+    IMPORT = 1
+    IMPROVE = 2
+
+    ACTION_TYPE_CHOICES = (
+        (IMPORT, "Import"),
+        (IMPROVE, "Improve"),
+    )
+
+    vulnerability = models.ForeignKey(
+        Vulnerability, on_delete=models.CASCADE, related_name="changelog"
+    )
+
+    action_type = models.PositiveSmallIntegerField(choices=ACTION_TYPE_CHOICES)
+
+    objects = VulnerabilityHistoryManager()
+
+    @classmethod
+    def log_import(cls, vulnerability, importer, source_url):
+        """
+        Creates History entry on Addition.
+        """
+        return cls.objects.log_action(
+            vulnerability=vulnerability,
+            action_type=VulnerabilityChangeLog.IMPORT,
+            actor_name=importer,
+            source_url=source_url,
+        )
+
+    @classmethod
+    def log_improve(cls, vulnerability, improver, source_url):
+        """
+        Creates History entry on Improvement.
+        """
+        return cls.objects.log_action(
+            vulnerability=vulnerability,
+            action_type=VulnerabilityChangeLog.IMPROVE,
+            actor_name=improver,
+            source_url=source_url,
+        )
+
+
+class PackageHistoryManager(models.Manager):
+    def get_for_object(self, package, **kwargs):
+        return self.filter(
+            package=package,
+            **kwargs,
+        )
+
+    def log_action(self, package, action_type, actor_name, source_url, related_vulnerability):
+        """
+        Creates a History entry for a given `obj` on Addition, Change, and Deletion.
+        We do not log addition for object that inherit the HistoryFieldsMixin since
+        the `created_by` and `created_date` are already set on its model.
+        """
+        return self.model.objects.get_or_create(
+            package=package,
+            action_type=action_type,
+            actor_name=actor_name,
+            source_url=source_url,
+            related_vulnerability=related_vulnerability,
+            software_version=VULNERABLECODE_VERSION,
+        )
+
+
+class PackageChangeLog(ChangeLog):
+
+    IMPORT = 1
+    AFFECTED_BY = 2
+    FIXING = 3
+
+    ACTION_TYPE_CHOICES = ((IMPORT, "Import"), (AFFECTED_BY, "Affected by"), (FIXING, "Fixing"))
+
+    package = models.ForeignKey(Package, on_delete=models.CASCADE, related_name="changelog")
+
+    # NOTES: We are not using foreign key because this is a log
+    # that we want to persist in case the VCID is not any more.
+
+    # TODO: We will change to foeign key in future once the models get stable
+    related_vulnerability = models.CharField(
+        max_length=1024,
+        null=True,
+        blank=True,
+        help_text="VCID of the vulnerability related to this package",
+    )
+
+    action_type = models.PositiveSmallIntegerField(choices=ACTION_TYPE_CHOICES)
+
+    objects = PackageHistoryManager()
+
+    @classmethod
+    def log_affected_by(cls, package, importer, source_url, related_vulnerability):
```

Comment:
Do we not want to keep Import history at package level?

Reply:
@keshav-space no, we just want to keep affected, fixed by or improvements on package, not the import logs.

Reply:
ack, in that case should we remove `IMPORT` as one of the action type choice in PackageChangeLog?

Reply:
Done!